### ISSUE TITLE
Dataset PUT Non-downloadable samples

### ISSUE THREAD

**Issue by jjeliaso** [AlexsLemonade/refinebio#2346]
The refine.bio api dataset PUT is giving a 400 response for "non_field_errors":["Non-downloadable sample(s)... in dataset"]}". This is causing our application to no longer be able to transfer to the refine.bio dataset download page.

For example, refine.bio dataset PUT and response:
PUT https://api.refine.bio/v1/dataset/c7f96f37-47a8-4a6c-aad7-d30aeb1d5068/ 400 (Bad Request)
_body: "{"non_field_errors":["Non-downloadable sample(s) 'SRR4101009, SRR6081902, SRR6081906, SRR6081907, SRR6081908, SRR6081909, SRR608893, SRR608894, SRR608895, SRR608896, SRR608897, SRR608898, SRR608899, SRR608900, SRR608901, SRR942849, SRR942850, SRR942851, SRR942852, SRR942853, SRR942854, SRR942855, SRR942856, SRR942857, SRR942858, SRR942859, SRR942860, SRR942861, SRR942862, SRR942863' in dataset"]}"

What is the preferred way to handle a refine.bio dataset PUT that contains samples either 'not found' or 'non-downloadable' from refine.bio?

**Comment by kurtwheeler** [comment]
So you cannot download a dataset that contains samples that we can't provide, either because we don't know about them or because we were not able to pre-process them. Therefore to handle that response, I would recommend removing those samples from the dataset, and then retrying your PUT.

However, if what you're really trying to do is just include all the samples for the experiment that we DO have available, then I would recommend is just providing a list containing the special keyword `"ALL"` for the experiment. i.e. `{"data": {"GSE123": ["ALL"]}}`. The endpoint will replace this with all the samples that are available for that experiment. If there's still none, then a 400 will be returned at which point I would recommend informing your user that refine.bio has no samples for that experiment available for download.

**Comment by jjeliaso** [comment]
Thanks for the suggestion, but we do not necessarily want all the samples from the refine.bio experiment. 

The dataset PUT 400 response does return a 'non downloadable' samples error message ("non_field_errors": "Non-downloadable sample(s)"). In a roundabout way I can use this error message to remove the samples from the our dataset, then run the dataset PUT again with only samples found at refine.bio. This works with two issues. First, the initial dataset PUT purposefully generates an error (the 400 response) to get data. Second, the 400 response "non_field_errors" value contains text not related to the list of samples (["Non-downloadable sample(s)...  in dataset"]), which requires additional string manipulation to remove. Also, parsing an error message to get data is not a robust solution as our code will break if the refine.bio error string ever changes.

Can you add a dataset PUT API route for us that accepts a dataset and returns the dataset 'non downloadable' samples? The purpose for this route is to return the samples 'not found' and 'non-downloadable' in a given dataset. This new route would work the same as the dataset PUT 400 'non downloadable' response, but return the string of 'non-downloadable' samples such as, { "body": { "non_downloadable_samples": "ERR3230071, ERR3230072, ERR3230073" }}.

**Comment by kurtwheeler** [comment]
Good point! I agree that expecting users to parse the text for those sample accessions is a bad design. Sorry about that, we should have thought of that when implementing it the first time. We'll change that response to have two keys: `message` and `non_downloadable_samples`. The message will say something like "There were non-downloadable samples in the dataset, see the non_downloadable_samples key for a list of sample accessions that are not available for download.".

About: 
> First, the initial dataset PUT purposefully generates an error (the 400 response) to get data.

I don't actually agree this is an issue. There's nothing wrong with getting HTTP error codes from an API. I think hitting an endpoint to test whether something exists or is valid is a common pattern for API driven programs. I think that with the above modifications to the 400 response, an additional route shouldn't be necessary.
